website/blog/unusual-builds-w-bytes.md

```diff
@@ -0,0 +1,217 @@
+After closer inspection of the cache statistics, we narrowed it down to a few key observations:
+
+1. There was little to no discrimination between the output artifacts being downloaded.
+Specifically, a large portion of what was being downloaded were our `go_test` binaries, which are never used because the test actions are always executed remotely in our CI setup.
+
+2. Most of the download requests included `prefetcher` metadata.
+Almost all gRPC requests made by Bazel include special metadata.
+
+```protobuf
+// An optional Metadata to attach to any RPC request to tell the server about an
+// external context of the request. The server may use this for logging or other
+// purposes. To use it, the client attaches the header to the call using the
+// canonical proto serialization:
+//
+// * name: `build.bazel.remote.execution.v2.requestmetadata-bin`
+// * contents: the base64 encoded binary `RequestMetadata` message.
+// ...
+message RequestMetadata {
+ ...
+
+ // An identifier that ties multiple requests to the same action.
+ // For example, multiple requests to the CAS, Action Cache, and Execution
+ // API are used in order to compile foo.cc.
+ string action_id = 2;
+
+ // An identifier that ties multiple actions together to a final result.
+ // For example, multiple actions are required to build and run foo_test.
+ string tool_invocation_id = 3;
+
+ ...
+
+ // A brief description of the kind of action, for example, CppCompile or GoLink.
+ // There is no standard agreed set of values for this, and they are expected to vary between different client tools.
+ string action_mnemonic = 5;
+
+ // An identifier for the target which produced this action.
+ // No guarantees are made around how many actions may relate to a single target.
+ string target_id = 6;
+
+ ...
+}
+```
+
+Here is an example of what this typically looks like in a normal request:
+```
+{
+   "action_id": "bd7fc99ea67ef952abe64d72dba270f04afe6424a1703ed12b3c935e13110597",
+   "tool_invocation_id": "a3b93bf2-8b16-4a72-b52f-636bc4db47a0",
+   "action_mnemonic": "GoLink",
+   "target_id": "//some/package/path:mypackage_test"
+}
+```
+
+However, there is a special class of request in Bazel called `prefetcher` requests.
+```
+{
+   "action_id": "prefetcher",
+   "tool_invocation_id": "8c2a7238-ce49-4149-8bfb-e7a6020294c2",
+   "action_mnemonic": "GoLink",
+   "target_id": "//some/package/path:mypackage_test"
+}
+```
+These are typically observed when an action is about to be executed locally, but some of its inputs were produced by a parent action that was executed remotely.
+In such cases, Bazel would typically issue "prefetch" download requests right after the parent action finishes so that the parent action's outputs can be downloaded to the local machine before the child action starts.
+These requests are often annotated with the `prefetcher` action ID and the target ID of the parent action.
```

Comment:
In the case of prefetched inputs, the ID should be the one of the consuming action, not the producing one. When fetching outputs, it's the producing action whose metadata is supplied.